xnet/tcp.go

```diff
@@ -0,0 +1,69 @@
+package xnet
+
+import (
+	"fmt"
+	"net"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// TCPSender is a Sender implementation that can write payload to the network
+// address and reuses TCP connections for the same addresses.
+type TCPSender struct {
+	Dialer net.Dialer
+
+	connections map[Address]net.Conn
+}
+
+// Send sends given payload to passed address. Data is sent using pool of TCP
+// connections. It returns number of bytes sent and error - if there was any.
+func (s *TCPSender) Send(addr Address, payload []byte) (int, error) {
+	if s.connections == nil {
+		s.connections = make(map[Address]net.Conn)
+	}
+	conn, ok := s.connections[addr]
+	if !ok {
+		newConn, err := s.dial(addr)
+		if err != nil {
+			return 0, fmt.Errorf("unable to dial %s address: %s", addr, err)
+		}
+		s.connections[addr] = newConn
+		conn = newConn
+	}
+	n, err := conn.Write(payload)
+	if err != nil {
+		// let's be nice and at least try to close connection on our side
+		closeErr := s.connections[addr].Close()
+		if closeErr != nil {
+			log.WithError(closeErr).Warn("Unable to close TCP connection properly")
```

Comment:
We will see this message before the actual error that might have caused it (above). Why not log the above message first too? Or return a MultiError.

Reply:
@AlfredBroda fixed